xmppserver/src/main/java/org/jivesoftware/openfire/disco/IQDiscoInfoHandler.java

```diff
@@ -505,15 +505,21 @@ private void restoreCacheContent() {
     public static DataForm getFirstDataForm(Set<DataForm> dataForms){
         if(dataForms != null && !dataForms.isEmpty()){
             Iterator<DataForm> iterator = dataForms.iterator();
-            if (iterator.hasNext()) {
-                Log.warn("Set Data List contains "+dataForms.size()+" DataForms."+
-                "Only the first one of the DataForms will be returned.");
+            DataForm dataForm = null;
```

Comment:
Although your test proves that the implementation is correct, it looks a bit. I think you could make it simpler, by splitting it in three parts:

```
if (dataForms.isEmpty() ) {
  return null;
}

if (dataForms.size() > 1) {
  Log.warn(" ... ");
}

return dataForms.iterator().next();
```

Reply:
I think it will not be complete when the collection has two elements the first null et the second not null !! it will be better to return the second instead to return null. the  function returns the first item that is not null

Reply:
imagine in another case if the collection has 4 elements and the no null form is in the fourth position

Reply:
Streams!

`return dataForms.stream().filter(Objects::nonNull).findAny().orElse(null);`

Also, it's a set, so un-order, so firndFirst/findLast would be non-deterministic anyway.

(Sorry, I'll go back in my hole now)

Reply:
Try what Greg said (although I think it's unlikely that there will be 'null' items in the collection in the first place.